File: crux_kv/shell/csharp/KeyValue.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Threading.Tasks;

/// <summary>
/// The shell side of <c>crux_kv</c>, as one method per operation.
/// </summary>
/// <remarks>
/// <para>
/// Each method's shape is the generated <c>IEffectHandler</c>'s, so an app's
/// handler holds an instance and delegates in one line:
/// </para>
/// <code>
/// public sealed class MyHandler : IEffectHandler
/// {
///     private readonly IKeyValueHandler _kv = new FileKeyValueHandler(directory);
///
///     public Task&lt;ValueResult&gt; KvGet(Get operation) => _kv.Get(operation);
///     public Task&lt;ValueResult&gt; KvSet(Set operation) => _kv.Set(operation);
/// }
/// </code>
/// <para>
/// Implement it yourself to put the store somewhere else, or to replace a
/// single method and leave the rest.
/// </para>
/// <para>
/// <c>Get</c> answers with what is stored under the key, <c>Value.None</c> when
/// nothing is. <c>Set</c> and <c>Delete</c> answer with the value they replaced
/// or removed, which is also <c>Value.None</c> when there was none. Nothing
/// here is an error unless the store itself failed; a missing key is an answer.
/// </para>
/// </remarks>
public interface IKeyValueHandler
{
    /// <summary>Read the bytes stored under <c>operation.Key</c>.</summary>
    Task<ValueResult> Get(Get operation);

    /// <summary>Write <c>operation.Value</c>, answering with the value it replaced.</summary>
    Task<ValueResult> Set(Set operation);

    /// <summary>Remove <c>operation.Key</c>, answering with the value it removed.</summary>
    Task<ValueResult> Delete(Delete operation);

    /// <summary>Whether <c>operation.Key</c> is in the store.</summary>
    Task<BoolResult> Exists(Exists operation);

    /// <summary>The keys starting with <c>operation.Prefix</c>, from <c>operation.Cursor</c>.</summary>
    Task<KeysResult> ListKeys(ListKeys operation);
}

/// <summary>
/// An <see cref="IKeyValueHandler"/> that keeps one file per key in a directory.
/// </summary>
/// <remarks>
/// <para>
/// The BCL has no key-value store of its own, and files are what every runtime
/// has: no database and no dependency, and a store an app can point at its
/// application-data folder, a temporary directory in a test, or anywhere else
/// it can write.
/// </para>
/// <para>
/// A key is percent-encoded to make its file name, so a key with a
/// <c>/</c> in it is one file and not a directory, and <c>ListKeys</c> can read
/// the keys back out of the directory listing.
/// </para>
/// <para>
/// A write is atomic and flushed to the device, so a value survives a crash
/// whole or not at all. Beyond that this is a plain file store: every read goes
/// to the filesystem, there is no cache, nothing spans more than one key, and
/// nothing observes a change. A shell that wants those conforms its own type to
/// <c>IKeyValueHandler</c> and provides that instead.
/// </para>
/// </remarks>
public sealed class FileKeyValueHandler : IKeyValueHandler
{
    private readonly string _directory;

    /// <summary>Keeps the files in <paramref name="directory"/>, created on first use.</summary>
    public FileKeyValueHandler(string directory)
    {
        _directory = directory;
    }

    public Task<ValueResult> Get(Get operation) =>
        Task.FromResult(Io(() => new ValueResult.Ok(Read(Path(operation.Key)))));

    public Task<ValueResult> Set(Set operation) =>
        Task.FromResult(Io(() =>
        {
            var path = Path(operation.Key);
            var previous = Read(path);
            Directory.CreateDirectory(_directory);
            var bytes = new byte[operation.Value.Count];
            operation.Value.CopyTo(bytes, 0);
            Write(path, bytes);
            return new ValueResult.Ok(previous);
        }));

    public Task<ValueResult> Delete(Delete operation) =>
        Task.FromResult(Io(() =>
        {
            var path = Path(operation.Key);
            var previous = Read(path);
            // `File.Delete` ignores a file that is not there, but throws
            // `DirectoryNotFoundException` when the directory holding it is
            // not there either, which a store nothing has been written to yet
            // is. Removing a key that was never set is an answer, not an
            // error, so the delete only happens when there is something to
            // delete.
            if (previous is not Value.None)
            {
                File.Delete(path);
            }
            return new ValueResult.Ok(previous);
        }));

    public Task<BoolResult> Exists(Exists operation)
    {
        try
        {
            return Task.FromResult<BoolResult>(new BoolResult.Ok(File.Exists(Path(operation.Key))));
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            return Task.FromResult<BoolResult>(new BoolResult.Err(new KeyValueError.Io(e.Message)));
        }
    }

    /// <summary>
    /// One page holds every remaining key, so the answer's cursor is always 0:
    /// the directory is listed in one go anyway, and paging it would only
    /// invite a caller to hold a cursor across a write.
    /// </summary>
    public Task<KeysResult> ListKeys(ListKeys operation)
    {
        List<string> keys;
        try
        {
            keys = new List<string>();
            if (Directory.Exists(_directory))
            {
                foreach (var file in Directory.EnumerateFiles(_directory))
                {
                    var key = Uri.UnescapeDataString(System.IO.Path.GetFileName(file));
                    if (key.StartsWith(operation.Prefix, StringComparison.Ordinal))
                    {
                        keys.Add(key);
                    }
                }
            }
            keys.Sort(StringComparer.Ordinal);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            return Task.FromResult<KeysResult>(new KeysResult.Err(new KeyValueError.Io(e.Message)));
        }

        if (operation.Cursor > (ulong)keys.Count)
        {
            return Task.FromResult<KeysResult>(new KeysResult.Err(new KeyValueError.CursorNotFound()));
        }

        var page = keys.GetRange((int)operation.Cursor, keys.Count - (int)operation.Cursor);
        return Task.FromResult<KeysResult>(new KeysResult.Ok(new KeyPage
        {
            Keys = new ObservableCollection<string>(page),
            NextCursor = 0,
        }));
    }

    private static Value Read(string path)
    {
        if (!File.Exists(path))
        {
            return new Value.None();
        }
        return new Value.Bytes(new ObservableCollection<byte>(File.ReadAllBytes(path)));
    }

    private string Path(string key) =>
        System.IO.Path.Combine(_directory, Uri.EscapeDataString(key));

    /// <summary>
    /// Writes <paramref name="bytes"/> to <paramref name="target"/> so that a
    /// reader sees either the whole of the new value or the whole of the old
    /// one.
    /// </summary>
    /// <remarks>
    /// The bytes go to a temporary file in a <c>tmp</c> directory inside the
    /// store, are flushed to the device, and are then moved over the target: a
    /// move within one volume is a rename, so a crash cannot leave a
    /// half-written value behind. The temporaries are in a directory of their
    /// own so that one left behind by a killed process is never mistaken for a
    /// key, since <c>ListKeys</c> reads the store's files and a directory is
    /// not one of them.
    /// </remarks>
    private void Write(string target, byte[] bytes)
    {
        var temporaries = System.IO.Path.Combine(_directory, "tmp");
        Directory.CreateDirectory(temporaries);
        var temporary = System.IO.Path.Combine(temporaries, System.IO.Path.GetRandomFileName());
        try
        {
            using (var stream = new FileStream(temporary, FileMode.CreateNew, FileAccess.Write))
            {
                stream.Write(bytes, 0, bytes.Length);
                stream.Flush(flushToDisk: true);
            }
            File.Move(temporary, target, overwrite: true);
        }
        finally
        {
            if (File.Exists(temporary))
            {
                File.Delete(temporary);
            }
        }
    }

    private static ValueResult Io(Func<ValueResult> block)
    {
        try
        {
            return block();
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            return new ValueResult.Err(new KeyValueError.Io(e.Message));
        }
    }
}

File: crux_kv/shell/kotlin/KeyValue.kt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException
import java.net.URLDecoder
import java.net.URLEncoder

/// The shell side of `crux_kv`, as one method per operation.
///
/// Each method's shape is the generated `EffectHandler`'s, so an app's handler
/// holds an instance and delegates in one line:
///
/// ```kotlin
/// class MyHandler(context: Context) : EffectHandler {
///     private val kv = FileKeyValueHandler(context.filesDir)
///
///     override suspend fun kvGet(operation: GetValue): ValueResult = kv.get(operation)
///     override suspend fun kvSet(operation: SetValue): ValueResult = kv.set(operation)
/// }
/// ```
///
/// Implement it yourself to put the store somewhere else — `DataStore`, a
/// database, a keystore — or to replace a single method and leave the rest.
///
/// ## What each operation answers with
///
/// `get` answers with what is stored under the key, `Value.None` when nothing
/// is. `set` and `delete` answer with the value they replaced or removed, which
/// is also `Value.None` when there was none. Nothing here is an error unless
/// the store itself failed; a missing key is an answer.
interface KeyValueHandler {
    /// Read the bytes stored under `operation.key`.
    suspend fun get(operation: GetValue): ValueResult

    /// Write `operation.value`, answering with the value it replaced.
    suspend fun set(operation: SetValue): ValueResult

    /// Remove `operation.key`, answering with the value it removed.
    suspend fun delete(operation: DeleteValue): ValueResult

    /// Whether `operation.key` is in the store.
    suspend fun exists(operation: KeyExists): ExistsResult

    /// The keys starting with `operation.prefix`, from `operation.cursor`.
    suspend fun listKeys(operation: ListKeys): KeysResult
}

/// A [KeyValueHandler] that keeps one file per key in a directory.
///
/// The JDK has no key-value store of its own, and files are what every JVM
/// has: no database, no `android.*`, and a store an app can point at
/// `context.filesDir`, a temporary directory in a test, or anywhere else it can
/// write.
///
/// A key is percent-encoded to make its file name, so a key with a `/` in it is
/// one file and not a directory, and `listKeys` can read the keys back out of
/// the directory listing.
///
/// ## What it gives you, and what it does not
///
/// A write is atomic and forced to the device, so a value survives a crash
/// whole or not at all. Beyond that this is a plain file store: every read
/// goes to the filesystem, there is no cache, nothing spans more than one key,
/// and nothing observes a change. A shell that wants those — `DataStore` or
/// Room on Android, a database anywhere — conforms its own type to
/// [KeyValueHandler] and provides that instead, which is a change to one
/// declaration.
///
/// @param directory where the files live; created on first use.
class FileKeyValueHandler(
    private val directory: File,
) : KeyValueHandler {
    override suspend fun get(operation: GetValue): ValueResult =
        withContext(Dispatchers.IO) {
            io { ValueResult.Ok(read(operation.key)) }
        }

    override suspend fun set(operation: SetValue): ValueResult =
        withContext(Dispatchers.IO) {
            io {
                val previous = read(operation.key)
                directory.mkdirs()
                write(file(operation.key), operation.value.map { it.toByte() }.toByteArray())
                ValueResult.Ok(previous)
            }
        }

    override suspend fun delete(operation: DeleteValue): ValueResult =
        withContext(Dispatchers.IO) {
            io {
                val previous = read(operation.key)
                file(operation.key).delete()
                ValueResult.Ok(previous)
            }
        }

    override suspend fun exists(operation: KeyExists): ExistsResult =
        withContext(Dispatchers.IO) {
            try {
                ExistsResult.Ok(file(operation.key).isFile)
            } catch (e: IOException) {
                ExistsResult.Err(KeyValueError.Io(e.message ?: "IO error"))
            } catch (e: SecurityException) {
                ExistsResult.Err(KeyValueError.Io(e.message ?: "access denied"))
            }
        }

    /// One page holds every remaining key, so the answer's cursor is always 0:
    /// the directory is listed in one go anyway, and paging it would only
    /// invite a caller to hold a cursor across a write.
    override suspend fun listKeys(operation: ListKeys): KeysResult =
        withContext(Dispatchers.IO) {
            val keys =
                try {
                    (directory.listFiles() ?: emptyArray())
                        .filter { it.isFile }
                        .map { URLDecoder.decode(it.name, Charsets.UTF_8.name()) }
                        .filter { it.startsWith(operation.prefix) }
                        .sorted()
                } catch (e: IOException) {
                    return@withContext KeysResult.Err(KeyValueError.Io(e.message ?: "IO error"))
                } catch (e: SecurityException) {
                    return@withContext KeysResult.Err(KeyValueError.Io(e.message ?: "access denied"))
                }

            if (operation.cursor > keys.size.toULong()) {
                return@withContext KeysResult.Err(KeyValueError.CursorNotFound)
            }

            KeysResult.Ok(KeyPage(keys.drop(operation.cursor.toInt()), 0uL))
        }

    private fun read(key: String): Value {
        val file = file(key)
        return if (file.isFile) Value.Bytes(file.readBytes().map { it.toUByte() }) else Value.None
    }

    private fun file(key: String): File = File(directory, URLEncoder.encode(key, Charsets.UTF_8.name()))

    /// Writes `bytes` to `target` so that a reader sees either the whole of
    /// the new value or the whole of the old one.
    ///
    /// The bytes go to a temporary file, are forced to the device, and are
    /// then renamed over the target. A rename within one filesystem is
    /// atomic, so a crash or a pulled battery leaves either the old value or
    /// the new one, never half of either.
    ///
    /// The temporary file is made in a `tmp` directory inside the store, so
    /// the rename stays within one filesystem — which is what `ATOMIC_MOVE`
    /// needs — and so that a temporary left behind by a killed process is
    /// never mistaken for a key: `listKeys` reads the files of the store
    /// directory, and a directory is not one of them.
    ///
    /// A filesystem is still free to refuse an atomic move, so the fallback is
    /// a plain replace — no longer atomic, and a crash during it can lose the
    /// value, but the alternative is failing a write that would otherwise
    /// succeed.
    private fun write(
        target: File,
        bytes: ByteArray,
    ) {
        val temporaries = File(directory, "tmp").apply { mkdirs() }
        val temporary = File.createTempFile("value", ".tmp", temporaries)
        try {
            java.io.FileOutputStream(temporary).use { out ->
                out.write(bytes)
                out.fd.sync()
            }
            try {
                java.nio.file.Files.move(
                    temporary.toPath(),
                    target.toPath(),
                    java.nio.file.StandardCopyOption.ATOMIC_MOVE,
                )
            } catch (_: java.nio.file.AtomicMoveNotSupportedException) {
                java.nio.file.Files.move(
                    temporary.toPath(),
                    target.toPath(),
                    java.nio.file.StandardCopyOption.REPLACE_EXISTING,
                )
            }
        } finally {
            temporary.delete()
        }
    }

    private inline fun io(block: () -> ValueResult): ValueResult =
        try {
            block()
        } catch (e: IOException) {
            ValueResult.Err(KeyValueError.Io(e.message ?: "IO error"))
        } catch (e: SecurityException) {
            ValueResult.Err(KeyValueError.Io(e.message ?: "access denied"))
        }
}
